mautic/mautic
# Contact reply to email campaign action

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This will complete and replace #4750 which seems to be abandoned and is looking for replies from the Marketer (monitors the SENT folder) so not from the contact. 

This PR refactors bounce, unsubscribe, and reply processing. Tests to come. 

To test this, you must have access to a non-Google or Yahoo email server as they will overwrite return paths. 

It also does some refactoring to Mautic's ability to remove configuration fields via security.php. Tests are included. Mainly load, change, and save the configuration to ensure that it works as expected. 

#### Steps to test this PR:
Bounces
1. Configure the Monitored Email section of the Email Settings in Mautic's configuration to successfully communicate with your IMAP server.
2. Select a folder for bounces and save
3. Create a new contact with LewisMMcKeon@teleworm.us as an email (teleworm is a domain to help test bounces)
4. Send an email to the contact and wait till you get the bounce notification
5. Run `php app/console mautic:email:fetch`
6. The contact should be marked as DNC bounced

Unsubscribes
1. Select a folder for unsubscribes and save
2. Create a contact with a real address and send an email to them from a segment or campaign
3. If the mail client doesn't show the unsubscribe email address, view the headers to get it. 
4. Send an email from the contact to that email address
5. Run `php app/console mautic:email:fetch`
6. The contact should be marked as unsubscribed

Replies
1. Select a folder for replies section and save
2. Create a campaign to send an email, then add a replies to email decision, then an action to tag the contact as "Replied"
3. Create a contact with a real address, add them to the campaign and trigger the campaign. 
4. Reply to the email as the contact
5. Run `php app/console mautic:email:fetch`
6. The contact should have an entry in their timeline for the campaign replied to email decision and also tagged as Replied

#### List deprecations along with the new alternative:
1. Mautic\EmailBundle\Helper\MessageHelper is deprecated. Use the proper InterfaceProcessor instead.
